ble/services/EnvironmentalService.h

```diff
@@ -0,0 +1,117 @@
+        ble(_ble),
+        temperatureCharacteristic(GattCharacteristic::UUID_TEMPERATURE_CHAR,
+                              (uint8_t *) &temperature,
+                              (temperature_en) ? 2 : 0, // minLength
+                              (temperature_en) ? 2 : 0, // maxLength
+                              GattCharacteristic::BLE_GATT_CHAR_PROPERTIES_READ),
+        humidityCharacteristic(GattCharacteristic::UUID_HUMIDITY_CHAR,
+                              (uint8_t *) &humidity,
+                              (humidity_en) ? 2 : 0, // minLength
+                              (humidity_en) ? 2 : 0, // maxLength
+                              GattCharacteristic::BLE_GATT_CHAR_PROPERTIES_READ),
+        pressureCharacteristic(GattCharacteristic::UUID_PRESSURE_CHAR,
+                              (uint8_t *) &pressure,
+                              (pressure_en) ? 4 : 0, // minLength
+                              (pressure_en) ? 4 : 0, // maxLength
+                              GattCharacteristic::BLE_GATT_CHAR_PROPERTIES_READ)
+        {
+        static bool serviceAdded = false; /* We should only ever need to add the information service once. */
+        if (serviceAdded) {
+            return;
+        }
+
+        GattCharacteristic *charTable[] = { &humidityCharacteristic,
+                                            &pressureCharacteristic,
+                                            &temperatureCharacteristic };
+
+        GattService environmentalService(GattService::UUID_ENVIRONMENTAL_SERVICE, charTable,
+                                                     sizeof(charTable) / sizeof(GattCharacteristic *));
+        
+        ble.gattServer().addService(environmentalService);
+        serviceAdded = true;
+    }
+
+    /**
+     * @brief   Update humidity characteristic.
+     * @param   newHumidityVal New humidity measurement.
+     */
+    void updateHumidity(uint16_t newHumidityVal)
+    {
+        humidity = (uint32_t) (newHumidityVal*100);                
+        ble.gattServer().write(humidityCharacteristic.getValueHandle(), (uint8_t *) &humidity, 2);
+    }
+
+    /**
+     * @brief   Update pressure characteristic.
+     * @param   newPressureVal New pressure measurement.
+     */
+    void updatePressure(uint32_t newPressureVal)
+    {
+        pressure = (uint32_t) (newPressureVal*10);        
+        ble.gattServer().write(pressureCharacteristic.getValueHandle(), (uint8_t *) &pressure, 4);
+    }
+
+    /**
+     * @brief   Update temperature characteristic.
+     * @param   newTemperatureVal New temperature measurement.
+     */
+    void updateTemperature(float newTemperatureVal)
+    {
+        temperature = (int16_t) (newTemperatureVal*100);
+        ble.gattServer().write(temperatureCharacteristic.getValueHandle(), (uint8_t *) &temperature, 2);
+    }
+
+private:
+    BLE &ble;
+
+    int16_t temperature;
+    GattCharacteristic temperatureCharacteristic;
```

Comment:
instead of using GattCharacteristic, could you try to use ReadOnlyGattCharacteristic<int16_t> as the type? refer to https://github.com/ARMmbed/ble/blob/master/ble/services/BatteryService.h#L63